### ISSUE TITLE
Zypper search should not acquire a lock

### ISSUE THREAD

**Issue by robinro** [openSUSE/zypper#108]
`zypper search` acquires a lock and thus prevents from running it twice simultaneously. Since this is (or should be) a read-only operation (at least when run with `--no-refresh`) no locking is needed.

In addition to being annoying for interactive use, this is really problematic for automated tools.
The zypper module of ansible uses `zypper search` to determine whether a package is already installed and thus can fail even in check-mode due to the search command failing because of locking.

**Comment by mlandres** [comment]
Unfortunately there's no quick fix for it. We need to introduce a  Readers-Writer lock to protect the searches from commands actually performing a refresh.

You (ansible) could set e.g. `ZYPP_LOCK_TIMEOUT=60` (seconds), to make libzypp wait longer for the lock becoming available. 

You can also catch zypper returning `7 - ZYPPER_EXIT_ZYPP_LOCKED`.and retry the search.

**Comment by robinro** [comment]
Thanks @mlandres for the quick response.

I was aware of `7 - ZYPPER_EXIT_ZYPP_LOCKED` and already had a PR to check for that in ansible, but `ZYPP_LOCK_TIMEOUT=60` is actually the nicer solution for our usecase.
Do I understand correctly that the default for this is `0`?

Maybe it's a good idea to hint to that variable in the error message that fails to acquire the lock. Or at least mention it on the zypper man page.

Another workaround for the locks: Run `zypper search` as non-root user while root hold a lock. Then that lock is ignored. Sadly thats no option for ansible.

**Comment by mlandres** [comment]
Yes, the default is `0` -  a single attempt to get the lock. 

A non-root user behaves slightly different, as he just consumes what's present in the caches, so he can't do much harm. Root will on the fly update missing/outdated cache entries, that's why some protection is needed.

**Comment by robinro** [comment]
In the code I just found `ZYPP_READONLY_HACK`, which does what I'm looking for and works in opensuse 13.2..42.2
How much of a "hack" is this? Can we have an official readonly flag?

**Comment by mlandres** [comment]
I know it's tempting, but this is not what you want. The hack was introduced for KIWI where we have a controlled environment (scripted and chrooted). It must not be used in a real system as it makes zypp vulnerable to race conditions. Using it YOU promise that you know no races will occur.

If it works for you, you should use `ZYPP_LOCK_TIMEOUT`. Zypper will be enhanced to allow more concurrency, but it's a somewhat bigger change and will need some time.

**Comment by robinro** [comment]
What's the worst that can happen, Zypper crashing or some corruption?
The `search` command only reads with that flag, right? And other commands (that have the lock) can read/write the cache/... in that time. So the worst might be a read on an inconsistent cache, which is a minor risc in my view. 

`ZYPP_LOCK_TIMEOUT` covers only a subset of the usecases, so I would prefer another solution.

**Comment by mlandres** [comment]
No, the flag has no influence on HOW the commands operate. It only omits acquiring the lock. 
The search command (as root) will still write caches or download files if it has to. 

At the time the hack was introduced, zypper/yast failed to indicate if '--root' was used, thus libzypp always placed the lock in the outer system, not in the chrooted one. So the hack was used to prevent locking the wrong system, not to enable concurrency in a single system.
